[sw, tock] Add the option to build Tock locally

jon-flatley opened the pull request:
Adds a -T flag to meson_init.sh that causes Tock to build from sw/device/tock/tock_local rather than using git dependencies. This is only a quality of life change for those who want to do local Tock development.

mcy commented on sw/device/tock/meson.build:
```diff
@@ -7,8 +7,9 @@ cargo = find_program('cargo')
 
-manifest_path = meson.source_root() / 'sw/device/tock/boards/opentitan/Cargo.toml'
-toolchain_file = meson.source_root() / 'sw/device/tock/boards/opentitan/rust-toolchain'
+board_dir = meson.source_root() / 'sw/device/tock/boards/opentitan'
```
-> `tock_board_dir`?

jon-flatley replied:
Done.

mcy commented on sw/device/tock/meson.build:
```diff
@@ -18,6 +19,12 @@ rust_flags = '-C link-arg=-Tlayout.ld ' + \
 
 build_tock_cmd = meson.source_root() / 'util/build_tock.sh'
 
+if get_option('tock_local')
```
I would suggest lifting this `get_option` to the root Meson file, like the rest of them.

jon-flatley replied:
Done.

jon-flatley commented:
New push:
- I didn't like that the `tock_local` folder was just a checked in empty folder, and I didn't like the nested git repos that it encouraged so I changed it to a `.gitignore`d symlink.
- I didn't like how the setup was manual, so I automated it and added it as part of the `-T` flag to `meson_init.sh`. Now `-T` takes an argument to the path of a Tock checkout.
- Local builds modified the `Cargo.lock` file so I added similar functionality to what is done for `Cargo.toml`. The difference here is that meson will produce a `Cargo_local.lock` that is ignored by git since there's no realistic scenario where this would need to be checked into version control. Depending on the build type meson will place a symlink to the correct file, as is done for `Cargo.toml`.

I'm still not a fan of how this requires allowing meson to touch files in the source directory. I don't see a good way of avoiding this with `Cargo.toml` and `Cargo.lock` since `cargo` flags don't provide a way to overload dependencies.

The other approach I could think of was to have two separate bin crates that just contain `Cargo.*` files with overloaded dependencies and effectively nothing else, so the dependency graph would look something like:

```
[Various Tock crates]
      \   |   /
      opentitan
     /         \
  local       remote
```

But this is overly complicated compared to two symlinks that get shuffled around, which is why the symlink approach is what is used in this PR.

silvestrst commented on meson_init.sh:
```diff
@@ -94,6 +100,26 @@ if [[ "${FLAGS_assert}" == true ]]; then
   fi
 fi
 
+TOCK_LOCAL=false
+if [[ ! -z "${ARG_tock_dir}" ]]; then
+  TOCK_LOCAL=true
+
+  if [[ ! -d "${ARG_tock_dir}" ]]; then
+    echo "Referenced Tock directory ${ARG_tock_dir} doesn't exist. Aborting." >&2
+    exit 1
+  fi
+
+  if [[ ! -f "${TOCK_SYMLINK}" || -L "${TOCK_SYMLINK}" ]]; then
+    echo "Creating symlink to Tock project at ${ARG_tock_dir}."
+    ln -sf "${ARG_tock_dir}" "${TOCK_SYMLINK}"
+  elif [[ -d "${TOCK_SYMLINK}" ]]; then
+    echo "Tock directory $TOCK_SYMLINK exists in place. Leaving as is."
+  else
+    echo "Placing Tock at ${TOCK_SYMLINK} would clobber something. Aborting." >&2
+    exit 1
+  fi
+fi
+
```
This probably could be split out in it's own `.sh `file that is called from `meson_init.sh`.

jon-flatley replied:
Fair point, no reason to add unnecessary clutter to `meson_init.sh`.

silvestrst commented:
I think somewhere this change must have a "neon sign" warning users that local copy has to be of the same version (d37ef3d2, currently I think). Otherwise you risk running into errors because of the mismatch between local and upstream opentitan versions (just experienced that). Only remembered it by comparing build logs and noticing that remote version pulls a specific Tock OS hash.

gkelly commented on sw/device/tock/boards/opentitan/Cargo_local.toml:
```diff
@@ -0,0 +1,26 @@
+[package]
```
Is there a way that this can be a symlink by default so that a person could just come in and `cargo build` without having to invoke meson?
